File: src/main/javascript/skynet-the-chasm.js

var R = parseInt(readline()); // the length of the road before the gap.
var G = parseInt(readline()); // the length of the gap.
var L = parseInt(readline()); // the length of the landing platform.
var neccessarySpeed = G+1;

// game loop
while (true) {
    var S = parseInt(readline()); // the motorbike's speed.
    var X = parseInt(readline()); // the position on the road of the motorbike.

    function shouldSlowDown() {
        function isOnPlatform() {
            return (X >= (R+G));
        }
        function isTooFast() {
            return (S > neccessarySpeed);
        }

        return isOnPlatform() || isTooFast();
    }

    function shouldJump() {
        function willReachPlatform() {
            return ((X+S) >= (R+G));
        }
        return willReachPlatform();
    }

    function shouldSpeedUp() {
        return (S < neccessarySpeed);
    }


    if (shouldSlowDown()) {
        print('SLOW');
    }
    else if (shouldJump()) {
        print('JUMP');
    }
    else if(shouldSpeedUp()) {
        print('SPEED');
    }
    else {
        print('WAIT');
    }
}

File: src/main/haskell/skynet-the-chasm.hs
import System.IO
import Control.Monad

main :: IO ()
main = do
    hSetBuffering stdout NoBuffering -- DO NOT REMOVE
    
    road <- fmap read getLine :: IO Int
    gap <- fmap read getLine :: IO Int
    platform <- fmap read getLine :: IO Int
    loop gap road

loop :: Int -> Int -> IO ()
loop gap road = do
    speed <- fmap read getLine :: IO Int
    coordx <- fmap read getLine :: IO Int
    
    putStrLn $ whatToDo coordx speed gap road
    
    loop gap road
    

targetSpeedForGap gap = gap + 1

shouldSlowDown x speed gap road =
  (isOnPlatform x gap road) || (isTooFast speed gap)
  where
    isOnPlatform x gap road = x >= (gap + road)
    isTooFast speed gap = speed > (targetSpeedForGap gap)
    
shouldSpeedUp speed gap = speed < (targetSpeedForGap gap)

shouldJump x speed gap road = (x + speed) >= (gap + road)

whatToDo x speed gap road 
  | shouldSlowDown x speed gap road = "SLOW"
  | shouldJump x speed gap road = "JUMP"
  | shouldSpeedUp speed gap = "SPEED"
  | otherwise = "WAIT"

